realtek: rtl931x: Add support for Plasma Cloud PSX28+ESX28 Switch

ecsv opened the pull request:
The Plasma Cloud ESX28 Switch is a 24 + 4 port multi-GBit switch with 24x 10/100/1000/2500BaseT Ethernet ports and 4x SFP+ module slot. The PSX28 uses the same hardware but also acts as POE++ PSE on all ethernet ports.

Hardware:

- RTL9312C SoC
- Macronix MX25L25645G (32MB flash)
- 512MB DDR3 SDRAM
- RTL8231 GPIO extender to control the port LEDs
- 6x RTL8224 4x 10m/100m/1/2.5 Gigabit PHY
- SFP+ 4x 10GBit slot
- RTL8239 POE++ PSE controller with frontend MCU (only PSX28)

The switch is powered directly via AC.

The external RS232 serial connector (RJ45, Cisco pinout) can be used to access the terminal. Serial connection is via 115200 baud, 8N1.

A reset button is accessible through a hole in the front panel.

Installation
------------

* The device can be flashed by using sysupgrade command. Either from the original vendor firmware or using an initramfs (see "Debug")
* Connect serial on front panel. Connection parameters: 115200 8N1
* The image (here `psx28`) must be copied using scp to /tmp of the device

      scp openwrt-realtek-rtl931x-plasmacloud_psx28-squashfs-sysupgrade.bin root@[IP address of the device]:/tmp/

* start sysupgrade without saving the original vendor configuration

      sysupgrade -n /tmp/openwrt-realtek-rtl931x-plasmacloud_psx28-squashfs-sysupgrade.bin

Installation via u-boot
-----------------------

If you have an TFTP server connected to the switch, it is possible to directly install the device using the factory image from u-boot

    # setup networking and IP of TFP server
    rtk network on
    setenv ipaddr 10.100.100.99
    setenv serverip 10.100.100.20
    
    # get factory image
    tftp 0x84000000 factory.bin
    
    # erase firmware partitions
    sf probe 0
    sf erase 0x5e0000 0x1a20000
    
    # write firmware to both partitions
    sf write ${fileaddr} 0x5e0000 ${filesize}
    sf write ${fileaddr} 0x12f0000 ${filesize}
    
    # adjust the boot commands
    setenv bootargs "mtdparts=spi0.0:768k(u-boot),64k(u-boot-env),64k(u-boot-env2),5120k(reserved),13376k(inactive),13376k(firmware2)"
    setenv bootcmd "rtk init; bootm 0xb52f0000"
    
    # restart
    reset

Debug
-----

* Connect serial on front panel. Connection parameters: 115200 8N1.
* A tftp server is required, tftpd-hpa works well.
* Power the device, at U-Boot start rapidly hit Esc key to stop autoboot
* Enter passwords: "1234" or "plasmapsx"
* Enable network:

      rtk network on

* Change ip address of device:

      setenv ipaddr 192.168.1.6

* Download initramfs from TFTP server:

      tftpboot 0x84000000 192.168.1.111:openwrt-realtek-rtl931x-plasmacloud_psx28-initramfs-kernel.bin

* Boot loaded file:

      bootm 0x84000000

---

The lag of proper SFP support is a known problem (for now) in OpenWrt RTL931x. See also #19579 (which was still using the old serdes code.

It is still a starting point for some RTL931x based switches in OpenWrt. And hopefully some work which is then based on it.

---

prerequisite-pr: #20263

ecsv committed:
realtek: rtl931x: Enable parsing of u-boot nvmem layouts

To be able to read out the ethaddr from the u-boot environment for MAC
address configuration, it is required to also enable the NVMEM layout
parsing code for the U-Boot env layout.

Signed-off-by: Sven Eckelmann <se@simonwunderlich.de>
Link: https://github.com/openwrt/openwrt/pull/20172
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

0xharshal committed:
realtek: rtl931x: Add support for Plasma Cloud PSX28 Switch

The Plasma Cloud PSX28 Switch is a 24 + 4 port multi-GBit switch with
24x 10/100/1000/2500BaseT Ethernet ports and 4x SFP+ module slot.

Hardware:

- RTL9312C SoC
- Macronix MX25L25645G (32MB flash)
- 512MB DDR3 SDRAM
- RTL8231 GPIO extender to control the port LEDs
- 6x RTL8224 4x 10m/100m/1/2.5 Gigabit PHY
- SFP+ 4x 10GBit slot
- RTL8239 POE++ PSE controller with frontend MCU

The switch is powered directly via AC.

The external RS232 serial connector (RJ45, Cisco pinout) can be used to
access the terminal. Serial connection is via 115200 baud, 8N1.

A reset button is accessible through a hole in the front panel.

Installation
------------

* The device can be flashed by using sysupgrade command. Either from the
  original vendor firmware or using an initramfs (see "Debug")
* Connect serial on front panel. Connection parameters: 115200 8N1
* The image must be copied using scp to /tmp of the device

      scp openwrt-realtek-rtl931x-plasmacloud_psx28-squashfs-sysupgrade.bin root@[IP address of the device]:/tmp/

* start sysupgrade without saving the original vendor configuration

      sysupgrade -n /tmp/openwrt-realtek-rtl931x-plasmacloud_psx28-squashfs-sysupgrade.bin

Installation via u-boot
-----------------------

If you have an TFTP server connected to the switch, it is possible to
directly install the device using the factory image from u-boot

    # setup networking and IP of TFP server
    rtk network on
    setenv ipaddr 10.100.100.99
    setenv serverip 10.100.100.20

    # get factory image
    tftp 0x84000000 factory.bin

    # erase firmware partitions
    sf probe 0
    sf erase 0x5e0000 0x1a20000

    # write firmware to both partitions
    sf write ${fileaddr} 0x5e0000 ${filesize}
    sf write ${fileaddr} 0x12f0000 ${filesize}

    # adjust the boot commands
    setenv bootargs "mtdparts=spi0.0:768k(u-boot),64k(u-boot-env),64k(u-boot-env2),5120k(reserved),13376k(inactive),13376k(firmware2)"
    setenv bootcmd "rtk init; bootm 0xb52f0000"

    # restart
    reset

Debug
-----

* Connect serial on front panel. Connection parameters: 115200 8N1.
* A tftp server is required, tftpd-hpa works well.
* Power the device, at U-Boot start rapidly hit Esc key to stop autoboot
* Enter passwords: "1234" or "plasmapsx"
* Enable network:

      rtk network on

* Change ip address of device:

      setenv ipaddr 192.168.1.6

* Download initramfs from TFTP server:

      tftpboot 0x84000000 192.168.1.111:openwrt-realtek-rtl931x-plasmacloud_psx28-initramfs-kernel.bin

* Boot loaded file:

      bootm 0x84000000

Signed-off-by: Harshal Gohel <hg@simonwunderlich.de>
Co-developed-by: Sven Eckelmann <se@simonwunderlich.de>
Signed-off-by: Sven Eckelmann <se@simonwunderlich.de>
Link: https://github.com/openwrt/openwrt/pull/20172
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

0xharshal committed:
realtek: rtl931x: Add support for Plasma Cloud ESX28 Switch

The Plasma Cloud ESX28 Switch is a 24 + 4 port multi-GBit switch with
24x 10/100/1000/2500BaseT Ethernet ports and 4x SFP+ module slot.

Hardware:

- RTL9312C SoC
- Macronix MX25L25645G (32MB flash)
- 512MB DDR3 SDRAM
- RTL8231 GPIO extender to control the port LEDs
- 6x RTL8224 4x 10m/100m/1/2.5 Gigabit PHY
- SFP+ 4x 10GBit slot

The switch is powered directly via AC.

The external RS232 serial connector (RJ45, Cisco pinout) can be used to
access the terminal. Serial connection is via 115200 baud, 8N1.

A reset button is accessible through a hole in the front panel.

Installation
------------

* The device can be flashed by using sysupgrade command. Either from the
  original vendor firmware or using an initramfs (see "Debug")
* Connect serial on front panel. Connection parameters: 115200 8N1
* The image must be copied using scp to /tmp of the device

      scp openwrt-realtek-rtl931x-plasmacloud_esx28-squashfs-sysupgrade.bin root@[IP address of the device]:/tmp/

* start sysupgrade without saving the original vendor configuration

      sysupgrade -n /tmp/openwrt-realtek-rtl931x-plasmacloud_esx28-squashfs-sysupgrade.bin

Installation via u-boot
-----------------------

If you have an TFTP server connected to the switch, it is possible to
directly install the device using the factory image from u-boot

    # setup networking and IP of TFP server
    rtk network on
    setenv ipaddr 10.100.100.99
    setenv serverip 10.100.100.20

    # get factory image
    tftp 0x84000000 factory.bin

    # erase firmware partitions
    sf probe 0
    sf erase 0x5e0000 0x1a20000

    # write firmware to both partitions
    sf write ${fileaddr} 0x5e0000 ${filesize}
    sf write ${fileaddr} 0x12f0000 ${filesize}

    # adjust the boot commands
    setenv bootargs "mtdparts=spi0.0:768k(u-boot),64k(u-boot-env),64k(u-boot-env2),5120k(reserved),13376k(inactive),13376k(firmware2)"
    setenv bootcmd "rtk init; bootm 0xb52f0000"

    # restart
    reset

Debug
-----

* Connect serial on front panel. Connection parameters: 115200 8N1.
* A tftp server is required, tftpd-hpa works well.
* Power the device, at U-Boot start rapidly hit Esc key to stop autoboot
* Enter passwords: "1234" or "plasmapsx"
* Enable network:

      rtk network on

* Change ip address of device:

      setenv ipaddr 192.168.1.6

* Download initramfs from TFTP server:

      tftpboot 0x84000000 192.168.1.111:openwrt-realtek-rtl931x-plasmacloud_esx28-initramfs-kernel.bin

* Boot loaded file:

      bootm 0x84000000

Signed-off-by: Harshal Gohel <hg@simonwunderlich.de>
Co-developed-by: Sven Eckelmann <se@simonwunderlich.de>
Signed-off-by: Sven Eckelmann <se@simonwunderlich.de>
Link: https://github.com/openwrt/openwrt/pull/20172
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

plappermaul commented:
Remove FEATURES+=source-only from openwrt/target/linux/realtek/rtl931x/target.mk in PSX28 commit.

plappermaul commented on target/linux/realtek/dts/rtl9312_plasmacloud_common.dtsi:
```diff
@@ -0,0 +1,610 @@
+		#size-cells = <0>;
+
+		/* copper ethernet */
+		SWITCH_PORT_SDS(0, 1, 2, usxgmii)
+		SWITCH_PORT_SDS(1, 2, 2, usxgmii)
+		SWITCH_PORT_SDS(4, 3, 2, usxgmii)
+		SWITCH_PORT_SDS(5, 4, 2, usxgmii)
+		SWITCH_PORT_SDS(8, 5, 3, usxgmii)
+		SWITCH_PORT_SDS(9, 6, 3, usxgmii)
```
According to kernel definition this cannot be USXGMII (1 port up to 10G). This is more likely 10g-qxgmii and with that phy we should start to make use of this. Especially to avoid confusion later when further converting SerDes. See https://elixir.bootlin.com/linux/v6.12.49/source/drivers/net/phy/phylink.c#L205 and https://elixir.bootlin.com/linux/v6.12.49/source/drivers/net/phy/phy-core.c#L107

If this gives issues when bringing up the link (had no time to test it yet) then we should add some workarounds with TODO comments.

ecsv replied:
I don't have all the HW for testing with me. I will have to postpone this to next week.

plappermaul replied:
Yes. Missed that on the PSX8/PSX10 review.

ecsv replied:
When device boots up, it will create an error:

```
rtl931x_phylink_mac_config: unknown serdes mode: 10g-qxgmii
```

See [rtl931x_phylink_mac_config](https://github.com/openwrt/openwrt/blob/2f6a467f6caf67f7bcba41adcaf6cd12cbdec612/target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/dsa.c#L669). `PHY_INTERFACE_MODE_10G_QXGMII` is not listed in there.

In the RTLSDK, you would have two different functions which handle the setup: https://github.com/plappermaul/realtek-doc/blob/7c3b2b297c81e8aa56bb22a11286310a7a7758a4/sources/rtk-xgs1210/src/dal/mango/dal_mango_construct.c#L1593

And they have quite some differences:


```diff
--- <unnamed>
+++ <unnamed>
@@ -1,85 +1,145 @@
-static int32 _dal_mango_construct_mac_usxgmii_10gs(uint32 unit, uint32 sds)
+static int32 _dal_mango_construct_mac_usxgmii_10gq(uint32 unit, uint32 sds)
 {
-    uint32  val = 1;
+    uint32  val;
     int32   ret;
 
+    val = 2;
+    RT_ERR_CHK(reg_array_field_write(unit, MANGO_USXGMII_SUBMODE_CTRLr,
+            REG_ARRAY_INDEX_NONE, sds, MANGO_USXGMII_SUBMODEf, &val), ret);
+
+    val = 1;
     switch (sds)
     {
         case 2:
-            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP0_1_CTRLr, MANGO_GROUP0_MAC_10G_ENf, &val), ret);
-            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP0_1_CTRLr, MANGO_GROUP0_MAC0_5G_ENf, &val), ret);
             RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP0_1_CTRLr, MANGO_GROUP0_MAC0_2P5G_ENf, &val), ret);
             RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP0_1_CTRLr, MANGO_GROUP0_MAC0_1G_ENf, &val), ret);
+            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP0_1_CTRLr, MANGO_GROUP0_MAC1_2P5G_ENf, &val), ret);
+            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP0_1_CTRLr, MANGO_GROUP0_MAC1_1G_ENf, &val), ret);
+            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP0_1_CTRLr, MANGO_GROUP0_MAC2_2P5G_ENf, &val), ret);
+            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP0_1_CTRLr, MANGO_GROUP0_MAC2_1G_ENf, &val), ret);
+            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP0_1_CTRLr, MANGO_GROUP0_MAC3_2P5G_ENf, &val), ret);
+            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP0_1_CTRLr, MANGO_GROUP0_MAC3_1G_ENf, &val), ret);
+
+            if (HWP_PORT_SDSID(unit, 4) == sds || HWP_PORT_SDSID(unit, 5) == sds)
+            {
+                val = 2;
+                RT_ERR_CHK(reg_field_write(unit, MANGO_PROT_SERDSE_MUX_CTRL_0r, MANGO_P4_GMII_SELf, &val), ret);
+                RT_ERR_CHK(reg_field_write(unit, MANGO_PROT_SERDSE_MUX_CTRL_1r, MANGO_SERDES2_2_GMII_SELf, &val), ret);
+                RT_ERR_CHK(reg_field_write(unit, MANGO_PROT_SERDSE_MUX_CTRL_0r, MANGO_P5_GMII_SELf, &val), ret);
+                RT_ERR_CHK(reg_field_write(unit, MANGO_PROT_SERDSE_MUX_CTRL_1r, MANGO_SERDES2_3_GMII_SELf, &val), ret);
+            }
             break;
         case 3:
-            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP0_1_CTRLr, MANGO_GROUP1_MAC_10G_ENf, &val), ret);
-            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP0_1_CTRLr, MANGO_GROUP1_MAC0_5G_ENf, &val), ret);
             RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP0_1_CTRLr, MANGO_GROUP1_MAC0_2P5G_ENf, &val), ret);
             RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP0_1_CTRLr, MANGO_GROUP1_MAC0_1G_ENf, &val), ret);
+            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP0_1_CTRLr, MANGO_GROUP1_MAC1_2P5G_ENf, &val), ret);
+            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP0_1_CTRLr, MANGO_GROUP1_MAC1_1G_ENf, &val), ret);
+            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP0_1_CTRLr, MANGO_GROUP1_MAC2_2P5G_ENf, &val), ret);
+            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP0_1_CTRLr, MANGO_GROUP1_MAC2_1G_ENf, &val), ret);
+            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP0_1_CTRLr, MANGO_GROUP1_MAC3_2P5G_ENf, &val), ret);
+            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP0_1_CTRLr, MANGO_GROUP1_MAC3_1G_ENf, &val), ret);
+
+            if (HWP_PORT_SDSID(unit, 12) == sds || HWP_PORT_SDSID(unit, 13) == sds)
+            {
+                val = 2;
+                RT_ERR_CHK(reg_field_write(unit, MANGO_PROT_SERDSE_MUX_CTRL_0r, MANGO_P12_GMII_SELf, &val), ret);
+                RT_ERR_CHK(reg_field_write(unit, MANGO_PROT_SERDSE_MUX_CTRL_1r, MANGO_SERDES3_2_GMII_SELf, &val), ret);
+                RT_ERR_CHK(reg_field_write(unit, MANGO_PROT_SERDSE_MUX_CTRL_0r, MANGO_P13_GMII_SELf, &val), ret);
+                RT_ERR_CHK(reg_field_write(unit, MANGO_PROT_SERDSE_MUX_CTRL_1r, MANGO_SERDES3_3_GMII_SELf, &val), ret);
+            }
             break;
         case 4:
-            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP2_3_CTRLr, MANGO_GROUP2_MAC_10G_ENf, &val), ret);
-            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP2_3_CTRLr, MANGO_GROUP2_MAC0_5G_ENf, &val), ret);
             RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP2_3_CTRLr, MANGO_GROUP2_MAC0_2P5G_ENf, &val), ret);
             RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP2_3_CTRLr, MANGO_GROUP2_MAC0_1G_ENf, &val), ret);
+            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP2_3_CTRLr, MANGO_GROUP2_MAC1_2P5G_ENf, &val), ret);
+            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP2_3_CTRLr, MANGO_GROUP2_MAC1_1G_ENf, &val), ret);
+            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP2_3_CTRLr, MANGO_GROUP2_MAC2_2P5G_ENf, &val), ret);
+            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP2_3_CTRLr, MANGO_GROUP2_MAC2_1G_ENf, &val), ret);
+            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP2_3_CTRLr, MANGO_GROUP2_MAC3_2P5G_ENf, &val), ret);
+            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP2_3_CTRLr, MANGO_GROUP2_MAC3_1G_ENf, &val), ret);
+
+            if (HWP_PORT_SDSID(unit, 20) == sds || HWP_PORT_SDSID(unit, 21) == sds)
+            {
+                val = 1;
+                RT_ERR_CHK(reg_field_write(unit, MANGO_PROT_SERDSE_MUX_CTRL_0r, MANGO_P20_GMII_SELf, &val), ret);
+                RT_ERR_CHK(reg_field_write(unit, MANGO_PROT_SERDSE_MUX_CTRL_1r, MANGO_SERDES4_2_GMII_SELf, &val), ret);
+                RT_ERR_CHK(reg_field_write(unit, MANGO_PROT_SERDSE_MUX_CTRL_0r, MANGO_P21_GMII_SELf, &val), ret);
+                RT_ERR_CHK(reg_field_write(unit, MANGO_PROT_SERDSE_MUX_CTRL_1r, MANGO_SERDES4_3_GMII_SELf, &val), ret);
+            }
             break;
         case 5:
-            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP2_3_CTRLr, MANGO_GROUP3_MAC_10G_ENf, &val), ret);
-            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP2_3_CTRLr, MANGO_GROUP3_MAC0_5G_ENf, &val), ret);
             RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP2_3_CTRLr, MANGO_GROUP3_MAC0_2P5G_ENf, &val), ret);
             RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP2_3_CTRLr, MANGO_GROUP3_MAC0_1G_ENf, &val), ret);
+            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP2_3_CTRLr, MANGO_GROUP3_MAC1_2P5G_ENf, &val), ret);
+            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP2_3_CTRLr, MANGO_GROUP3_MAC1_1G_ENf, &val), ret);
+            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP2_3_CTRLr, MANGO_GROUP3_MAC2_2P5G_ENf, &val), ret);
+            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP2_3_CTRLr, MANGO_GROUP3_MAC2_1G_ENf, &val), ret);
+            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP2_3_CTRLr, MANGO_GROUP3_MAC3_2P5G_ENf, &val), ret);
+            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP2_3_CTRLr, MANGO_GROUP3_MAC3_1G_ENf, &val), ret);
+
+            if (HWP_PORT_SDSID(unit, 28) == sds || HWP_PORT_SDSID(unit, 29) == sds)
+            {
+                val = 1;
+                RT_ERR_CHK(reg_field_write(unit, MANGO_PROT_SERDSE_MUX_CTRL_0r, MANGO_P28_GMII_SELf, &val), ret);
+                RT_ERR_CHK(reg_field_write(unit, MANGO_PROT_SERDSE_MUX_CTRL_1r, MANGO_SERDES5_2_GMII_SELf, &val), ret);
+                RT_ERR_CHK(reg_field_write(unit, MANGO_PROT_SERDSE_MUX_CTRL_0r, MANGO_P29_GMII_SELf, &val), ret);
+                RT_ERR_CHK(reg_field_write(unit, MANGO_PROT_SERDSE_MUX_CTRL_1r, MANGO_SERDES5_3_GMII_SELf, &val), ret);
+            }
             break;
         case 6:
-            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP4_CTRLr, MANGO_GROUP4_MAC_10G_ENf, &val), ret);
-            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP4_CTRLr, MANGO_GROUP4_MAC0_5G_ENf, &val), ret);
             RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP4_CTRLr, MANGO_GROUP4_MAC0_2P5G_ENf, &val), ret);
             RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP4_CTRLr, MANGO_GROUP4_MAC0_1G_ENf, &val), ret);
+            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP4_CTRLr, MANGO_GROUP4_MAC1_2P5G_ENf, &val), ret);
+            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP4_CTRLr, MANGO_GROUP4_MAC1_1G_ENf, &val), ret);
+            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP4_CTRLr, MANGO_GROUP4_MAC2_2P5G_ENf, &val), ret);
+            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP4_CTRLr, MANGO_GROUP4_MAC2_1G_ENf, &val), ret);
+            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP4_CTRLr, MANGO_GROUP4_MAC3_2P5G_ENf, &val), ret);
+            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP4_CTRLr, MANGO_GROUP4_MAC3_1G_ENf, &val), ret);
+
+            if (HWP_PORT_SDSID(unit, 36) == sds || HWP_PORT_SDSID(unit, 37) == sds)
+            {
+                val = 1;
+                RT_ERR_CHK(reg_field_write(unit, MANGO_PROT_SERDSE_MUX_CTRL_0r, MANGO_P36_GMII_SELf, &val), ret);
+                RT_ERR_CHK(reg_field_write(unit, MANGO_PROT_SERDSE_MUX_CTRL_1r, MANGO_SERDES6_2_GMII_SELf, &val), ret);
+                RT_ERR_CHK(reg_field_write(unit, MANGO_PROT_SERDSE_MUX_CTRL_0r, MANGO_P37_GMII_SELf, &val), ret);
+                RT_ERR_CHK(reg_field_write(unit, MANGO_PROT_SERDSE_MUX_CTRL_1r, MANGO_SERDES6_3_GMII_SELf, &val), ret);
+            }
             break;
         case 7:
-            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP5_CTRLr, MANGO_GROUP5_MAC_10G_ENf, &val), ret);
-            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP5_CTRLr, MANGO_GROUP5_MAC0_5G_ENf, &val), ret);
             RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP5_CTRLr, MANGO_GROUP5_MAC0_2P5G_ENf, &val), ret);
             RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP5_CTRLr, MANGO_GROUP5_MAC0_1G_ENf, &val), ret);
-            break;
-        case 8:
-            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP6_7_CTRLr, MANGO_GROUP6_MAC_10G_ENf, &val), ret);
-            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP6_7_CTRLr, MANGO_GROUP6_MAC0_5G_ENf, &val), ret);
-            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP6_7_CTRLr, MANGO_GROUP6_MAC0_2P5G_ENf, &val), ret);
-            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP6_7_CTRLr, MANGO_GROUP6_MAC0_1G_ENf, &val), ret);
+            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP5_CTRLr, MANGO_GROUP5_MAC1_2P5G_ENf, &val), ret);
+            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP5_CTRLr, MANGO_GROUP5_MAC1_1G_ENf, &val), ret);
+            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP5_CTRLr, MANGO_GROUP5_MAC2_2P5G_ENf, &val), ret);
+            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP5_CTRLr, MANGO_GROUP5_MAC2_1G_ENf, &val), ret);
+            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP5_CTRLr, MANGO_GROUP5_MAC3_2P5G_ENf, &val), ret);
+            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP5_CTRLr, MANGO_GROUP5_MAC3_1G_ENf, &val), ret);
+
+            if (HWP_PORT_SDSID(unit, 44) == sds || HWP_PORT_SDSID(unit, 45) == sds)
+            {
+                val = 1;
+                RT_ERR_CHK(reg_field_write(unit, MANGO_PROT_SERDSE_MUX_CTRL_0r, MANGO_P44_GMII_SELf, &val), ret);
+                RT_ERR_CHK(reg_field_write(unit, MANGO_PROT_SERDSE_MUX_CTRL_1r, MANGO_SERDES7_2_GMII_SELf, &val), ret);
+                RT_ERR_CHK(reg_field_write(unit, MANGO_PROT_SERDSE_MUX_CTRL_0r, MANGO_P45_GMII_SELf, &val), ret);
+                RT_ERR_CHK(reg_field_write(unit, MANGO_PROT_SERDSE_MUX_CTRL_1r, MANGO_SERDES7_3_GMII_SELf, &val), ret);
+            }
             break;
         case 9:
-            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP6_7_CTRLr, MANGO_GROUP7_MAC_10G_ENf, &val), ret);
-            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP6_7_CTRLr, MANGO_GROUP7_MAC0_5G_ENf, &val), ret);
-            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP6_7_CTRLr, MANGO_GROUP7_MAC0_2P5G_ENf, &val), ret);
-            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP6_7_CTRLr, MANGO_GROUP7_MAC0_1G_ENf, &val), ret);
-            break;
-        case 10:
-            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP8_11_CTRLr, MANGO_GROUP8_MAC_10G_ENf, &val), ret);
-            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP8_11_CTRLr, MANGO_GROUP8_MAC_5G_ENf, &val), ret);
-            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP8_11_CTRLr, MANGO_GROUP8_MAC_2P5G_ENf, &val), ret);
-            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP8_11_CTRLr, MANGO_GROUP8_MAC_1G_ENf, &val), ret);
-            break;
-        case 11:
-            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP8_11_CTRLr, MANGO_GROUP9_MAC_10G_ENf, &val), ret);
-            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP8_11_CTRLr, MANGO_GROUP9_MAC_5G_ENf, &val), ret);
-            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP8_11_CTRLr, MANGO_GROUP9_MAC_2P5G_ENf, &val), ret);
-            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP8_11_CTRLr, MANGO_GROUP9_MAC_1G_ENf, &val), ret);
-            break;
-        case 12:
-            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP8_11_CTRLr, MANGO_GROUP10_MAC_10G_ENf, &val), ret);
-            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP8_11_CTRLr, MANGO_GROUP10_MAC_5G_ENf, &val), ret);
-            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP8_11_CTRLr, MANGO_GROUP10_MAC_2P5G_ENf, &val), ret);
-            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP8_11_CTRLr, MANGO_GROUP10_MAC_1G_ENf, &val), ret);
-            break;
-        case 13:
-            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP8_11_CTRLr, MANGO_GROUP11_MAC_10G_ENf, &val), ret);
-            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP8_11_CTRLr, MANGO_GROUP11_MAC_5G_ENf, &val), ret);
-            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP8_11_CTRLr, MANGO_GROUP11_MAC_2P5G_ENf, &val), ret);
-            RT_ERR_CHK(reg_field_write(unit, MANGO_MAC_GROUP8_11_CTRLr, MANGO_GROUP11_MAC_1G_ENf, &val), ret);
+            if (HWP_PORT_SDSID(unit, 42) == sds || HWP_PORT_SDSID(unit, 43) == sds ||
+                    HWP_PORT_SDSID(unit, 46) == sds || HWP_PORT_SDSID(unit, 47) == sds)
+            {
+                val = 1;
+                RT_ERR_CHK(reg_field_write(unit, MANGO_PROT_SERDSE_MUX_CTRL_0r, MANGO_P42_GMII_SELf, &val), ret);
+                RT_ERR_CHK(reg_field_write(unit, MANGO_PROT_SERDSE_MUX_CTRL_0r, MANGO_P43_GMII_SELf, &val), ret);
+                RT_ERR_CHK(reg_field_write(unit, MANGO_PROT_SERDSE_MUX_CTRL_0r, MANGO_P46_GMII_SELf, &val), ret);
+                RT_ERR_CHK(reg_field_write(unit, MANGO_PROT_SERDSE_MUX_CTRL_0r, MANGO_P47_GMII_SELf, &val), ret);
+                RT_ERR_CHK(reg_field_write(unit, MANGO_PROT_SERDSE_MUX_CTRL_1r, MANGO_P50_GMII_SELf, &val), ret);
+                RT_ERR_CHK(reg_field_write(unit, MANGO_PROT_SERDSE_MUX_CTRL_1r, MANGO_SERDES9_0_GMII_SELf, &val), ret);
+            }
             break;
         default:
-            return RT_ERR_FAILED;
+            return RT_ERR_OK;
     }
 
     return ret;
-}   /* end of _dal_mango_construct_mac_usxgmii_10gs */
+}   /* end of _dal_mango_construct_mac_usxgmii_10gq */

```


But if you look at `rtl931x_sds_init` in OpenWrt, you see suddenly following comment:

```
case PHY_INTERFACE_MODE_USXGMII: /* MII_USXGMII_10GSXGMII/10GDXGMII/10GQXGMII: */
```

I would therefore guess that @0xharshal made (based on this comment) the right decision to select `usxgmii` for his initial integration.


And to get it just working with the correct PHY mode string and the keep the rest of the setup the same, following change might be necessary:

```diff
From: Sven Eckelmann <se@simonwunderlich.de>
Date: Mon, 29 Sep 2025 16:07:54 +0200
Subject: realtek: Work around missing 10g-qxmgmii PHY mode

The current SerDes implementation for RTL931x handles 10G-QXGMII via the
"usxgmii" PHY mode. This is not 100% correct because it is not a single
port with 10G (max) but 4 ports with 2.5G each.

To allow setting of the "10g-qxgmii" phy mode, just change the code for now
to use the same codepaths as USXGMII. This has to be cleaned up further
during the SerDes driver rewrites.

Signed-off-by: Sven Eckelmann <se@simonwunderlich.de>

diff --git a/target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/common.c b/target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/common.c
index 6bae35633d36c86fb318defe3b5ae8850d16e424..d3a9b3f0d9c8888b23f18627ac412f6356a38abd 100644
--- a/target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/common.c
+++ b/target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/common.c
@@ -361,6 +361,8 @@ static int __init rtl83xx_mdio_probe(struct rtl838x_switch_priv *priv)
 			interface = PHY_INTERFACE_MODE_NA;
 		if (interface == PHY_INTERFACE_MODE_USXGMII)
 			priv->ports[pn].is2G5 = priv->ports[pn].is10G = true;
+		if (interface == PHY_INTERFACE_MODE_10G_QXGMII)
+			priv->ports[pn].is2G5 = true;
 		if (interface == PHY_INTERFACE_MODE_10GBASER)
 			priv->ports[pn].is10G = true;
 
diff --git a/target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/dsa.c b/target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/dsa.c
index eee3479aea961dca7ce9686d9e6c70eea24c329a..de99164c8d5ae7ccafe3165c8b9f64e9c354d55a 100644
--- a/target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/dsa.c
+++ b/target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/dsa.c
@@ -625,6 +625,7 @@ static void rtldsa_phylink_get_caps(struct dsa_switch *ds, int port,
 	config->mac_capabilities = MAC_ASYM_PAUSE | MAC_SYM_PAUSE | MAC_10 | MAC_100 |
 				   MAC_1000FD | MAC_2500FD | MAC_5000FD | MAC_10000FD;
 
+	__set_bit(PHY_INTERFACE_MODE_10G_QXGMII, config->supported_interfaces);
 	__set_bit(PHY_INTERFACE_MODE_1000BASEX, config->supported_interfaces);
 	__set_bit(PHY_INTERFACE_MODE_10GBASER, config->supported_interfaces);
 	__set_bit(PHY_INTERFACE_MODE_2500BASEX, config->supported_interfaces);
@@ -678,6 +679,10 @@ static void rtl931x_phylink_mac_config(struct dsa_switch *ds, int port,
 	pr_info("%s: speed %d sds_num %d\n", __func__, state->speed, sds_num);
 
 	switch (state->interface) {
+	case PHY_INTERFACE_MODE_10G_QXGMII:
+		band = rtl931x_sds_cmu_band_get(sds_num, PHY_INTERFACE_MODE_10G_QXGMII);
+		rtl931x_sds_init(sds_num, PHY_INTERFACE_MODE_10G_QXGMII);
+		break;
 	case PHY_INTERFACE_MODE_1000BASEX:
 		band = rtl931x_sds_cmu_band_get(sds_num, PHY_INTERFACE_MODE_1000BASEX);
 		rtl931x_sds_init(sds_num, PHY_INTERFACE_MODE_1000BASEX);
diff --git a/target/linux/realtek/files-6.12/drivers/net/mdio/mdio-realtek-otto.c b/target/linux/realtek/files-6.12/drivers/net/mdio/mdio-realtek-otto.c
index 0f55bf51f48fddcdba1a4d14658dc4470978aaf8..af4a0f6e47b11ca27c9739b41c6800dc5ee93c0f 100644
--- a/target/linux/realtek/files-6.12/drivers/net/mdio/mdio-realtek-otto.c
+++ b/target/linux/realtek/files-6.12/drivers/net/mdio/mdio-realtek-otto.c
@@ -1295,6 +1295,7 @@ static int rtmdio_930x_reset(struct mii_bus *bus)
 		switch (priv->interfaces[i]) {
 		case PHY_INTERFACE_MODE_10GBASER:
 			break;			/* Serdes: Value = 0 */
+		case PHY_INTERFACE_MODE_10G_QXGMII:
 		case PHY_INTERFACE_MODE_USXGMII:
 			v |= BIT(mac_type_bit[i]);
 			uses_usxgmii = true;
diff --git a/target/linux/realtek/files-6.12/drivers/net/phy/rtl83xx-phy.c b/target/linux/realtek/files-6.12/drivers/net/phy/rtl83xx-phy.c
index 9ea73e6fa18f0f2902c47b044c5749b7b953d7dc..6f9002d5dd838e4fe20bcaa0dd6da15592913cad 100644
--- a/target/linux/realtek/files-6.12/drivers/net/phy/rtl83xx-phy.c
+++ b/target/linux/realtek/files-6.12/drivers/net/phy/rtl83xx-phy.c
@@ -1413,6 +1413,7 @@ static void rtl9300_sds_tx_config(int sds, phy_interface_t phy_if)
 		pre_en = 0;
 		page = 0x29;
 		break;
+	case PHY_INTERFACE_MODE_10G_QXGMII:
 	case PHY_INTERFACE_MODE_10GBASER:
 	case PHY_INTERFACE_MODE_USXGMII:
 	case PHY_INTERFACE_MODE_XGMII:
@@ -2723,6 +2724,7 @@ static void rtl931x_sds_fiber_mode_set(u32 sds, phy_interface_t mode)
 		break; */
 
 
+	case PHY_INTERFACE_MODE_10G_QXGMII:
 	case PHY_INTERFACE_MODE_USXGMII:
 		val = 0x1B;
 		break;
@@ -2771,6 +2773,7 @@ static void rtl931x_cmu_type_set(u32 sds, phy_interface_t mode, int chiptype)
 
 	switch (mode) {
 	case PHY_INTERFACE_MODE_NA:
+	case PHY_INTERFACE_MODE_10G_QXGMII:
 	case PHY_INTERFACE_MODE_10GKR:
 	case PHY_INTERFACE_MODE_XGMII:
 	case PHY_INTERFACE_MODE_10GBASER:
@@ -2887,6 +2890,7 @@ static void rtl931x_sds_mii_mode_set(u32 sds, phy_interface_t mode)
 	case PHY_INTERFACE_MODE_XGMII:
 		val = 0x10; /* serdes mode XSGMII */
 		break;
+	case PHY_INTERFACE_MODE_10G_QXGMII:
 	case PHY_INTERFACE_MODE_USXGMII:
 	case PHY_INTERFACE_MODE_2500BASEX:
 		val = 0xD;
@@ -3000,6 +3004,7 @@ void rtl931x_sds_init(u32 sds, phy_interface_t mode)
 		break;
 
 	case PHY_INTERFACE_MODE_USXGMII: /* MII_USXGMII_10GSXGMII/10GDXGMII/10GQXGMII: */
+	case PHY_INTERFACE_MODE_10G_QXGMII:
 		u32 op_code = 0x6003;
 		u32 evenSds;
 
@@ -3104,7 +3109,8 @@ void rtl931x_sds_init(u32 sds, phy_interface_t mode)
 	sw_w32(val, RTL931X_PS_SERDES_OFF_MODE_CTRL_ADDR);
 	pr_debug("%s: RTL931X_PS_SERDES_OFF_MODE_CTRL_ADDR 0x%08X\n", __func__, sw_r32(RTL931X_PS_SERDES_OFF_MODE_CTRL_ADDR));
 
-	if (mode == PHY_INTERFACE_MODE_XGMII ||
+	if (mode == PHY_INTERFACE_MODE_10G_QXGMII ||
+	    mode == PHY_INTERFACE_MODE_XGMII ||
 	    mode == PHY_INTERFACE_MODE_QSGMII ||
 	    mode == PHY_INTERFACE_MODE_SGMII ||
 	    mode == PHY_INTERFACE_MODE_USXGMII) {

```


Yes, I can basically hear to you scream in agony - but this is really just a workaround (by a person not knowing much about the RTL931x SerDes) to get the correct PHY mode in the DTS. It is added for now in the PR

plappermaul commented:
Please drop the 10g-qxgmii patch immediately. That makes my brain hurt.

Afterwards first things first. Make pre-PSX28-fixup-PR (reuse the new PSX8/10 PR)

1. Make a commit for the RTL931x REALTEK_PHY kernel config.

2. Make a commit with a simple 10g-qxgmii to usxgmi conversion in dsa and mdio driver like this

```
of_get_phy_mode(dn, &interface))

if (interface == PHY_INTERFACE_MODE_10G_QXGMII) { <--- ADD
  interface = PHY_INTERFACE_MODE_USXGMII;
  dev_warn(..., "handle mode 10g-qsxgmii internally as usxgmii for now\n");
}

```

3. Third commit should be the dts port mode chage for the PSX8/10

Thanks.

ecsv commented:
Marked as ready again after #20239 was merged

@plappermaul, your change requests https://github.com/openwrt/openwrt/pull/20172#discussion_r2383986235 and https://github.com/openwrt/openwrt/pull/20172#discussion_r2383978783 should be resolved, correct?

plappermaul commented on target/linux/realtek/dts/rtl931x.dtsi:
```diff
@@ -296,6 +296,31 @@
 		};
 	};
 
+	pinmux@1b00103c {
+		compatible = "pinctrl-single";
+		reg = <0x1b00103c 0x4>;
+
+		pinctrl-single,bit-per-mux;
+		pinctrl-single,register-width = <32>;
+		pinctrl-single,function-mask = <0x1>;
+		#pinctrl-cells = <2>;
+
+		/* Enable GPIO 8, 9, 10 */
+		pinmux_disable_spi0: disable-spi0 {
+			pinctrl-single,bits = <0x0 0x0 0x400>;
+		};
+
+		/* Enable GPIO 12 */
+		pinmux_disable_spi0_cs1: disable-spi-cs1 {
+			pinctrl-single,bits = <0x0 0x0 0x200>;
+		};
+
+		/* Enable GPIO 11 */
+		pinmux_disable_spi0_cs0: disable-spi-cs0 {
+			pinctrl-single,bits = <0x0 0x0 0x100>;
+		};
+	};
+
```
Make this a separate precondition PR and explain what it is about. Especially how it is (not) related to the NOR and NAND controllers in the SoC.

ecsv replied:
Done, see #20263. Will rebase and de-draft this PR after #20263 was merged.

plappermaul commented on target/linux/realtek/rtl931x/target.mk:
```diff
@@ -4,7 +4,6 @@ SUBTARGET:=rtl931x
 CPU_TYPE:=24kc
 BOARD:=realtek
 BOARDNAME:=Realtek MIPS RTL931X
-FEATURES+=source-only
```
Make the "drop source-only" a separate final commit in this series.

plappermaul commented on target/linux/realtek/base-files/etc/init.d/hwmon_fancontrol:
```diff
@@ -0,0 +1,40 @@
+
+	case $(board_name) in
+	plasmacloud,psx28)
+		path_to_hwmon="$(find /sys/devices/platform/1b000000.switchcore/1b000000.switchcore:i2c@100c/i2c-4/4-002e/hwmon/ -type d -maxdepth 1 -mindepth 1 -name 'hwmon*')"
+
+		# adt7476 fan control chip
+		# 3 temp sensors. Set to lowest speed at 35C and max speed at 70C
+		# set critical temp to 80 as per the datasheet
+		echo 7 > "$path_to_hwmon/pwm2_auto_channels_temp"
```
Make this coding a device specific subfunction to keep the boot() function clean.

plappermaul commented on target/linux/realtek/dts/rtl9312_plasmacloud_psx28.dts:
```diff
@@ -0,0 +1,24 @@
+#include "rtl9312_plasmacloud_common.dtsi"
+
+/ {
+	compatible = "plasmacloud,psx28", "realtek,rtl838x-soc";
```
Please align to upstream and use rtl9312-soc. See https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/arch/mips/boot/dts/realtek?h=v6.17

ecsv committed:
realtek: drop source-only from rtl931x target

Now the rtl931x target has real devices that need to be built. Remove the
source-only flag to make the images available.

Signed-off-by: Sven Eckelmann <se@simonwunderlich.de>
Link: https://github.com/openwrt/openwrt/pull/20172
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

ecsv commented on target/linux/realtek/dts/rtl9312_plasmacloud_common.dtsi:
```diff
@@ -0,0 +1,612 @@
+&mdio_aux {
+	status = "okay";
+
+	gpio1: expander@24 {
+		compatible = "realtek,rtl8231";
+		reg = <24>;
```
Ehrm, that means that the mdio_aux bus address 24 (PA5) is used here. The auxilary mdio bus is using GPIO 1 (MDC/AM25) and GPIO 2 (MDIO/AL25) pins for this. They are switched using register 0x1b0007d4 (EXT_GPIO_GLB_CTRL) bit 8 (0 == GPIO, 1 == EXT_GPIO)

I don't know why GPIO 24 would be used here. Unlike SPI, MDIO is using inline addressing of attached peripherals. Basically, bit 4-8 of the MDIO header are the PHY address. And we selected here 24. It is more similar to I2C in this regard.

See also:

* https://github.com/openwrt/openwrt/blob/50c51504fd38704d6b669390dd0ddee1e3af704c/target/linux/realtek/patches-6.12/723-net-mdio-Add-Realtek-Otto-auxiliary-controller.patch#L157
* https://www.kernel.org/doc/Documentation/devicetree/bindings/net/mdio.yaml

You could say, the reg is not relevant for the expander itself - it is the information for the parent device (which provides the bus and needs to know where on the bus this expander would be - simply to perform the correct writes/reads).

ecsv commented:
Last rebase was just to get #20263 from  the main branch. No other change was performed (or at least intended). 

The PR was therefore also marked again as ready

plappermaul commented on target/linux/realtek/dts/rtl9312_plasmacloud_common.dtsi:
```diff
@@ -0,0 +1,612 @@
+		label-mac-device = &ethernet0;
+	};
+
+	memory@0 {
+		reg = <0x00000000 0x10000000>, /* low mem 256 MiB */
+		      <0x90000000 0x10000000>; /* high mem 256 MiB */
+	};
```
Are you sure that this works? @jonasjelonek can you confirm with your RTL931x?

Why am I asking. I always have issues with bringing up my RTL9311 (revision B) device with highmem. And it basically boils down to the fact "highmem on -> reatlek otto timer interrupts do not fire". Here some debug output.

```
Timer setup
[    0.000000] irq-mips-gic: gic_unmask_irq(1)
[    0.000000] irq-mips-gic: gic_set_affinity set affinitiy of hwirq 8 irq 8 to cpu 0
[    0.000000] timer_rtl_otto: ------------- 0 b8003210 rttm_setup_timer
[    0.000000] timer_rtl_otto: Before acknowledge b800321c result is 00000000
[    0.000000] timer_rtl_otto: After acknowledge b800321c result is 00000000
[    0.000000] irq-mips-gic: gic_unmask_irq(2)
[    0.000000] irq-mips-gic: gic_set_affinity set affinitiy of hwirq 9 irq 9 to cpu 0
[    0.000000] timer_rtl_otto: ------------- 0 b8003220 rttm_setup_timer
[    0.000000] timer_rtl_otto: Before acknowledge b800322c result is 00000000
[    0.000000] timer_rtl_otto: After acknowledge b800322c result is 00000000
[    0.000000] irq-mips-gic: gic_unmask_irq(3)
[    0.000000] irq-mips-gic: gic_set_affinity set affinitiy of hwirq 10 irq 10 to cpu 0
[    0.000000] timer_rtl_otto: ------------- 0 b8003230 rttm_setup_timer
[    0.000000] timer_rtl_otto: Before acknowledge b800323c result is 00000000
[    0.000000] timer_rtl_otto: After acknowledge b800323c result is 00000000
[    0.000000] irq-mips-gic: gic_unmask_irq(4)
[    0.000000] irq-mips-gic: gic_set_affinity set affinitiy of hwirq 11 irq 11 to cpu 0
[    0.000000] timer_rtl_otto: ------------- 0 b8003240 rttm_setup_timer
[    0.000000] timer_rtl_otto: Before acknowledge b800324c result is 00000000
[    0.000000] timer_rtl_otto: After acknowledge b800324c result is 00000000
[    0.000000] timer_rtl_otto: ------------- 0 b8003250 rttm_setup_timer
[    0.000000] timer_rtl_otto: Before acknowledge b800325c result is 00000000
[    0.000000] timer_rtl_otto: After acknowledge b800325c result is 00000000
[    0.000000] clocksource: realtek_otto_timer: mask: 0xfffffff max_cycles: 0xfffffff, max_idle_ns: 38225208801 ns
[    0.000003] sched_clock: 28 bits at 3125kHz, resolution 320ns, wraps every 42949672800ns
[    0.009040] timer_rtl_otto: ------------- 0 b8003210 rttm_cpu_starting
[    0.016319] timer_rtl_otto: I enabled interrupt at b800321c result is 00100000
[    0.024381] timer_rtl_otto: ------------- 0 b8003210 rttm_state_shutdown
[    0.031850] timer_rtl_otto: Before acknowledge b800321c result is 00100000
[    0.039515] timer_rtl_otto: After acknowledge b800321c result is 00100000

Start timer and hope that interrupts fire periodically
[    0.047089] timer_rtl_otto: ------------- 0 b8003210 rttm_state_periodic
[    0.054564] timer_rtl_otto: wait for timer start
[    0.059712] timer_rtl_otto: timer started 00007a12 00003ed4 11000040 00100000
[    0.067927] START_KERNEL 40 on CPU 0
[    0.071951] START_KERNEL 41 on CPU 0
[    0.075939] START_KERNEL 42 on CPU 0
[    0.079945] START_KERNEL 43 on CPU 0
[    0.084330] START_KERNEL 44 on CPU 0
[    0.088349] START_KERNEL 45 on CPU 0
[    0.092342] START_KERNEL 46 on CPU 0
[    0.096447] START_KERNEL 47 on CPU 0
[    0.100466] START_KERNEL 48 on CPU 0
[    0.104450] START_KERNEL 49 on CPU 0
[    0.108428] START_KERNEL 50 on CPU 0
[    0.112444] START_KERNEL 51 on CPU 0
[    0.116427] START_KERNEL 52 on CPU 0
[    0.120405] START_KERNEL 53 on CPU 0
[    0.124383] START_KERNEL 54 on CPU 0
[    0.128376] START_KERNEL 55 on CPU 0
[    0.132365] CALIBRATE_DELAY 1 CPU 0
[    0.136255] CALIBRATE_DELAY 7 CPU 0
[    0.140143] Calibrating delay loop...

Hang in calibrate_delay_converge() because no interrupts pop up.
[    0.140149] CALIBRATE_DELAY_CONVERGE 1 CPU 0
[    0.149067] TIMER STATUS b8003210 : 00007a12, 000036f4, 11000040, 00110000

```

ecsv replied:
Yes, I was running this configuration for a while and it works fine.

jonasjelonek replied:
> I have one here and can test a DTS / PR. Just started collecting the needed information some days ago.

The branch I'm working on for this device is currently local-only but it would be no problem to push to my fork.

I could do a PR for this but it would be a draft and probably not ready soon. While I got working most side stuff, networking doesn't really work. Either I'm just too stupid, or it's because proper SFP support is still missing for RTL931x. Only the first port (which is initialized by the bootloader during the special ramboot procedure) actually works. Similar issues for two other RTL9313 devices I have here. 

Not sure what other opinions on this are, but I'm unsure if I should just open PRs for the devices I have even though not all of the stuff works. I just could argue for it to show that this is being worked on.

plappermaul reviewed (approved):
Sad to have missed that chance after all the hard work ... I hereby grant the right to activate the first RTL931x devices.

robimarko commented on target/linux/realtek/base-files/etc/init.d/hwmon_fancontrol:
Why not simply use thermal zones and cooling devices via DT?
